### PR TITLE
Add Chemdner dataset loader

### PR DESCRIPTION
**Pull request by qanastek**
Please name your PR after the issue it closes. You can use the following line: "Closes #ISSUE-NUMBER" where you replace the ISSUE-NUMBER with the one corresponding to your dataset.

If the following information is NOT present in the issue, please populate:

- **Name:** *name of the dataset*
- **Description:** *short description of the dataset (or link to social media or blog post)*
- **Paper:** *link to the dataset paper if available*
- **Data:** *link to the online home of the dataset*

### Checkbox

- [x] Confirm that this PR is linked to the dataset issue.
- [x] Create the dataloader script `biodatasets/my_dataset/my_dataset.py` (please use only lowercase and underscore for dataset naming).
- [x] Provide values for the `_CITATION`, `_DATASETNAME`, `_DESCRIPTION`, `_HOMEPAGE`, `_LICENSE`, `_URLs`, `_SUPPORTED_TASKS`, `_SOURCE_VERSION`, and `_BIGBIO_VERSION` variables.
- [x] Implement `_info()`, `_split_generators()` and `_generate_examples()` in dataloader script.
- [x] Make sure that the `BUILDER_CONFIGS` class attribute is a list with at least one `BigBioConfig` for the source schema and one for a bigbio schema.
- [x] Confirm dataloader script works with `datasets.load_dataset` function.
- [x] Confirm that your dataloader script passes the test suite run with `python -m tests.test_bigbio biodatasets/my_dataset/my_dataset.py`.

### CODE REVIEW

**Comment by qanastek** [comment]
Issue https://github.com/bigscience-workshop/biomedical/issues/28

File: biodatasets/chemdner/chemdner.py [hunk 1]
```diff
@@ -0,0 +1,394 @@
...
+entity mention (SACEM) class: abbreviation, family, formula, identifier, multiple, systematic and trivial.
+"""
+
+_DATASETNAME = "CHEMDNER"
+
+_HOMEPAGE = "https://biocreative.bioinformatics.udel.edu/resources/biocreative-iv/chemdner-corpus/"
+_LICENSE = "Unknown"
+
+_URLs = {
+    "source": "https://biocreative.bioinformatics.udel.edu/media/store/files/2014/chemdner_corpus.tar.gz",
```
[the remark this answers is not in the document]
Comment (qanastek) [reply biodatasets/chemdner/chemdner.py:131]: Ok, I will try!

File: biodatasets/chemdner/chemdner.py [hunk 2]
```diff
@@ -0,0 +1,394 @@
...
+            for a in p.annotations:
+
+                if a.text is None or len(a.text) <= 0:
+                    continue
+
+                a_type = a.infons.get("class")
+                # a_type = a.infons.get("type")
+
+                # no in-text annotation: only for document indexing
+                if a_type in ["MeSH_Indexing_Chemical", "OTHER"]:
```
[the remark this answers is not in the document]
Comment (qanastek) [reply biodatasets/chemdner/chemdner.py:287]: So, like this :

```python
# no in-text annotation: only for document indexing
if self.config.schema == "bigbio_kb" and a_type in ["MeSH_Indexing_Chemical", "OTHER"]:
    continue
```

[the remark this answers is not in the document]
Comment (qanastek) [reply biodatasets/chemdner/chemdner.py:287]: The data distribution :

With Mesh:

- train =>  37154
- test  =>  31553
- dev   =>  37189

Without Mesh:

- train => 29462
- test  => 25346
- dev   => 29523

**Comment by qanastek** [comment]
Commit 526e7d9c6a3ca1eb9491a169c74675c13a526cb5

File: biodatasets/chemdner/chemdner.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (sg-wbi) [reply biodatasets/chemdner/chemdner.py]: I saw this too. Thank you for pointing this out. There are indeed many "MESH Indexing" annotations.

[the remark this answers is not in the document]
Comment (sg-wbi) [reply biodatasets/chemdner/chemdner.py]: Minor cosmetic thing: it seems "chemdner" does not have annotations with "OTHER" type

**Comment by sg-wbi** [comment]
Thank you for implementing the changes! Everything looks fine. You made me realize that we could use the "MeSH_Indexing_Chemical " as labels for a `text_classification` schema. I asked internally about this and get back to you asap.

File: biodatasets/chemdner/chemdner.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (qanastek) [reply biodatasets/chemdner/chemdner.py]: I saw, no labels "OTHER"are available but, do you want me to remove it from the condition ?

[the remark this answers is not in the document]
Comment (sg-wbi) [reply biodatasets/chemdner/chemdner.py]: Yep thank you. As I said, minor cosmetic: no sense to check for something we know it's not there.

**Comment by qanastek** [comment]
Commit 6251db84a154dac6894c5ea3a693eb8a2630b519

File: biodatasets/chemdner/chemdner.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (qanastek) [reply biodatasets/chemdner/chemdner.py]: Done in the last Commit 6251db84a154dac6894c5ea3a693eb8a2630b519

**Comment by galtay** [comment]
@sg-wbi thanks for self assigning this one! I'll jump to some of the other PRs

**Comment by galtay** [comment]
just a quick note, we have some extra files in this PR besides just the `chemdner` loader

**Comment by qanastek** [comment]
Fixed I remove other files

**Comment by sg-wbi** [comment]
Thanks galtey! There is an updated on the schema for `text_classification` which allows for mulitple labels. See the "#general" channel in Discord. 

@qanastek Could you please add this view to the dataset? You can use this PR #348 as an example.

Then we can close this!

**Comment by qanastek** [comment]
During the tests of the Text Classification task, I crashed during the `datasets.load_dataset()` function. Have you any ideas ? I use your code base.

The error :

```plain
======================================================================
ERROR: runTest (__main__.TestDataLoader)
Run all tests that check:
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/d/Projects/LIA/huggingface/BigScience BioMedical/pull_resquests/Final_ChemdNER/biomedical/tests/test_bigbio.py", line 164, in setUp
    self.datasets_bigbio[schema] = datasets.load_dataset(
  File "/home/home/miniconda3/envs/bigscience-biomedical/lib/python3.10/site-packages/datasets/load.py", line 1702, in load_dataset
    builder_instance.download_and_prepare(
  File "/home/home/miniconda3/envs/bigscience-biomedical/lib/python3.10/site-packages/datasets/builder.py", line 594, in download_and_prepare
    self._download_and_prepare(
  File "/home/home/miniconda3/envs/bigscience-biomedical/lib/python3.10/site-packages/datasets/builder.py", line 683, in _download_and_prepare
    self._prepare_split(split_generator, **prepare_split_kwargs)
  File "/home/home/miniconda3/envs/bigscience-biomedical/lib/python3.10/site-packages/datasets/builder.py", line 1080, in _prepare_split
    example = self.info.features.encode_example(record)
  File "/home/home/miniconda3/envs/bigscience-biomedical/lib/python3.10/site-packages/datasets/features/features.py", line 1287, in encode_example
    return encode_nested_example(self, example)
  File "/home/home/miniconda3/envs/bigscience-biomedical/lib/python3.10/site-packages/datasets/features/features.py", line 971, in encode_nested_example
    return {
  File "/home/home/miniconda3/envs/bigscience-biomedical/lib/python3.10/site-packages/datasets/features/features.py", line 971, in <dictcomp>
    return {
  File "/home/home/miniconda3/envs/bigscience-biomedical/lib/python3.10/site-packages/datasets/utils/py_utils.py", line 152, in zip_dict
    yield key, tuple(d[key] for d in dicts)
  File "/home/home/miniconda3/envs/bigscience-biomedical/lib/python3.10/site-packages/datasets/utils/py_utils.py", line 152, in <genexpr>
    yield key, tuple(d[key] for d in dicts)
KeyError: 'label'

----------------------------------------------------------------------
Ran 1 test in 15.950s

FAILED (errors=1)

```

My code :

```python
    def _get_textcls_example(self, d: bioc.BioCDocument) -> Dict:

        example = {"document_id": d.id, "text": [], "labels": []}

        for p in d.passages:

            example["text"].append(p.text)

            for a in p.annotations:

                if a.infons.get("type") == "MeSH_Indexing_Chemical":

                    example["labels"].append(a.infons.get("identifier"))

        example["text"] = " ".join(example["text"])

        return example
```

The `datasets.load_dataset` is searching for a `label` field, except that we are giving it a `labels` one.

Have you passed the tests with the loading script of `nlmchem` ?

**Comment by sg-wbi** [comment]
The test for `nlmchem` passes. Are you working on the latest version of the repo?
```bash
git pull upstream master
```

**Comment by qanastek** [comment]
Commit 540901f261e0f818d199e961ce3f753bfc398718

**Comment by sg-wbi** [comment]
Thank you very much for implementing the additional view! One last thing and then it should be good to go, for real this time. If I run the tests I get these warnings:
"""
WARNING:__main__:Example:11379 - entity:11371  text:`None` != text_by_offset:``
Example:17134 - entity:17129  text:`None` != text_by_offset:``
Example:21504 - entity:21503  text:`None` != text_by_offset:``
Example:23227 - entity:23213  text:`None` != text_by_offset:``
"""
It seems that somehow some entities without text get into the dataset. Could you please check again your script and fix this?
Thank you!

**Comment by qanastek** [comment]
Nice job! I haven't seemed them since the output is very compact.

It was due to `None` presents in the data : 

- `{'id': '21503', 'type': 'Chemical', 'text': [None], 'offsets': [[1607, 1607]], 'normalized': []}`
- `{'id': '11371', 'type': 'Chemical', 'text': [None], 'offsets': [[179, 179]], 'normalized': []}`
- `{'id': '17129', 'type': 'Chemical', 'text': [None], 'offsets': [[55, 55]], 'normalized': []}`

I fixed it using a condition : 

https://github.com/bigscience-workshop/biomedical/blob/ac8fb5ec13cc8f8d40b876d623fa364fbce7118e/biodatasets/chemdner/chemdner.py#L293-L294

I hope it's ok!

**Comment by sg-wbi** [comment]
@qanastek

**Comment by qanastek** [comment]
Like this ?

```python
if self.config.schema == "bigbio_kb" and a_type == "MeSH_Indexing_Chemical":
    continue

if (a.text == None or a.text == "") and self.config.schema == "bigbio_kb":
    continue
```

**Comment by sg-wbi** [comment]
The first one is easier to read

**Comment by qanastek** [comment]
Commit 9cf91005fa444747b717633a42ad672ff56df1e5

**Comment by sg-wbi** [comment]
Great! Thank you for bearing all the changes, especially w/ the schema change. Thanks again for your contribution!

**Comment by qanastek** [comment]
@sg-wbi Can you switch the issue (https://github.com/bigscience-workshop/biomedical/issues/28) status to `Done` to keep track of the advancement ?

**Comment by qanastek** [comment]
I think so, like Gabriel do on the Issue https://github.com/bigscience-workshop/biomedical/issues/53

**Comment by sg-wbi** [comment]
Now?

**Comment by qanastek** [comment]
Perfect!

**Pull request merged**